### PR TITLE
chore: move local dependency to packages

### LINKED ISSUE #23333: Clean up client workspaces
Move yarn env to client folder

### PR DESCRIPTION
**Pull request by KelvinOm**
## Description
1. Move everything related to client from app folder to client folder (`.yarn`, `yarn.lock`, package.json, .gitignore)
2. Move `ast` and `rst` to client packages
3. Fix running scripts in packages
4. Add running unit tests in packages in CI

TODO: It is necessary to consider enabling the `nmHoistingLimits: workspaces` option, since now all packages are hoisted to the root, there may be issues with dependencies in workspaces. Also, there is a possibility of implicit use of packages.

https://yarnpkg.com/configuration/yarnrc#nmHoistingLimits

#### PR fixes following issue(s)
Fixes #23333

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

#### How Has This Been Tested?
- [x] Manual
- [x] Jest
- [x] Cypress

#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### CODE REVIEW

**Comment by KelvinOm** [comment]
> @KelvinOm , Can you find out what the convention is on where to put the builds of individual workspace packages.
> We are putting it in workspace directory here, should the builds be put in root, i.e client?

@SatishGandham If there is a need to build something, then the artifacts should be at the workspace level as if it were an isolated package. In any case, the point of workspaces is not to build anything except the main application.

File: app/client/package.json [hunk 1]
```diff
@@ -4,9 +4,13 @@
   "private": true,
   "engines": {
     "node": "^16.14.0",
-    "npm": "^8.5.5"
+    "yarn": "^3.5.1"
```
Comment (iamakulov) [inline app/client/package.json:7]: 😍

File: app/client/packages/rts/build.sh [hunk 2]
```diff
@@ -7,8 +7,6 @@ rm -rf dist/
 # This is required for the first time build as node_modules is not present in the image
 yarn install --immutable
 yarn tsc && yarn tsc-alias
-# Keep copy of all dependencies in node_modules_bkp
-mv node_modules node_modules_bkp
```
Comment (iamakulov) [inline app/client/packages/rts/build.sh:11]: ❤️‍🔥

**Review by iamakulov — DISMISSED** [review]
Love this cleanup!

> TODO: It is necessary to consider enabling the `nmHoistingLimits: workspaces` option, since now all packages are hoisted to the root, there may be issues with dependencies in workspaces. Also, there is a possibility of implicit use of packages.

Not sure what the performance implications of this are (might be minuscule, might be significant). But just in case, if you’re only worried about some environments (eg CI builds) or some packages, it’s also possible to set that flag per command, with the `YARN_NM_HOISTING_LIMITS` variable. E.g.:

```sh
YARN_NM_HOISTING_LIMITS=workspaces yarn install
```

(This is [a convention](https://yarnpkg.com/configuration/yarnrc) that works for all config flags.)

**Comment by KelvinOm** [comment]
> Love this cleanup!
> 
> > TODO: It is necessary to consider enabling the `nmHoistingLimits: workspaces` option, since now all packages are hoisted to the root, there may be issues with dependencies in workspaces. Also, there is a possibility of implicit use of packages.
> 
> Not sure what the performance implications of this are (might be minuscule, might be significant). But just in case, if you’re only worried about some environments (eg CI builds) or some packages, it’s also possible to set that flag per command, with the `YARN_NM_HOISTING_LIMITS` variable. E.g.:
> 
> ```shell
> YARN_NM_HOISTING_LIMITS=workspaces yarn install
> ```
> 
> (This is [a convention](https://yarnpkg.com/configuration/yarnrc) that works for all config flags.)

@iamakulov Implicit use of dependencies can lead to unexpected consequences. Besides, this may get worse when we use workspaces more widely. 

As for performance, I still think that there is nothing faster than pnpm. At least we should try using the `nodeLinker: pnpm` option.

**Comment by KelvinOm** [comment]
[build for 8fc4f9f](https://github.com/appsmithorg/appsmith/actions/runs/5016976616)

**Comment by KelvinOm** [comment]
[build for e941933](https://github.com/appsmithorg/appsmith/actions/runs/5043720342)

**Comment by KelvinOm** [comment]
[build for bced199](https://github.com/appsmithorg/appsmith/actions/runs/5043863959)

**Comment by KelvinOm** [comment]
[build for 672a417](https://github.com/appsmithorg/appsmith/actions/runs/5043933055)

**Comment by KelvinOm** [comment]
[build for 7ea4a1f](https://github.com/appsmithorg/appsmith/actions/runs/5044264947)

**Pull request merged**